examples/pytorch/graphsage/train_cv_multi_gpu.py

```diff
@@ -0,0 +1,416 @@
+    """
+    Evaluate the model on the validation set specified by ``val_mask``.
+    g : The entire graph.
+    inputs : The features of all the nodes.
+    labels : The labels of all the nodes.
+    val_mask : A 0-1 mask indicating which nodes do we actually compute the accuracy for.
+    batch_size : Number of nodes to compute at the same time.
+    device : The GPU device to evaluate on.
+    """
+    model.eval()
+    with th.no_grad():
+        inputs = g.ndata['features']
+        pred, _ = model.inference(g, inputs, batch_size, device)
+    model.train()
+    return compute_acc(pred[val_mask], labels[val_mask])
+
+def load_subtensor(g, labels, blocks, dev_id):
+    """
+    Copys features and labels of a set of nodes onto GPU.
+    """
+    blocks[0].srcdata['features'] = g.ndata['features'][blocks[0].srcdata[dgl.NID]].to(dev_id)
+    blocks[-1].dstdata['label'] = labels[blocks[-1].dstdata[dgl.NID]].to(dev_id)
+    for i, block in enumerate(blocks):
+        hist_col = 'features' if i == 0 else 'hist_%d' % i
+        block.srcdata['hist'] = g.ndata[hist_col][block.srcdata[dgl.NID]].to(dev_id)
+        block.dstdata['agg_hist'] = g.ndata['sum_hist_%d' % i][block.dstdata[dgl.NID]].to(dev_id)
+        block.dstdata['agg_hist'] /= g.ndata['deg'][block.dstdata[dgl.NID]].to(dev_id)[:, None]
+
+def init_history(g, model, dev_id):
+    with th.no_grad():
+        history = model.inference(g, g.ndata['features'], 1000, dev_id)[1]
+        for layer in range(args.num_layers + 1):
+            if layer > 0:
+                hist_col = 'hist_%d' % layer
+                g.ndata['hist_%d' % layer] = history[layer - 1]
+            else:
+                hist_col = 'features'
+            agg_hist_col = 'sum_hist_%d' % layer
+            # TODO: can we avoid this column?
+            g.ndata['hist_delta_%d' % layer] = g.ndata[hist_col].clone().zero_()
+            g.update_all(fn.copy_u(hist_col, 'm'), fn.sum('m', agg_hist_col))
+
+        g.ndata['deg'] = g.in_degrees().float().clamp(min=1)    # avoid 0 in denominator
+
+def update_history(g, blocks):
+    with th.no_grad():
+        for i, block in enumerate(blocks):
+            ids = block.dstdata[dgl.NID]
+            hist_col = 'hist_%d' % (i + 1)
+            hist_delta_col = 'hist_delta_%d' % (i + 1)
+            agg_hist_col = 'sum_hist_%d' % (i + 1)
+            agg_new_col = 'agg_new_%d' % (i + 1)
+
+            h_new = block.dstdata['h_new'].cpu()
+            old_hist = g.ndata[hist_col][ids]
+            g.ndata[hist_delta_col][ids] = h_new - old_hist
+            g.ndata[hist_col][ids] = h_new
+            # TODO: can we avoid creating a new feature called agg_new_col?
+            g.push(
+                ids,
+                fn.copy_u(hist_delta_col, 'm'),
+                fn.sum('m', agg_new_col),
+                lambda nodes: {agg_hist_col: nodes.data[agg_hist_col] + nodes.data[agg_new_col]},
+                inplace=True)
```

Comment:
it seems you need to keep both history node embeddings for each layer as well as aggregated history for each layer. Is it right?

Reply:
I think it's correct. After every batch, both history and aggregate of history will be update in `update_history` function.

But actually this is not very efficient, because updating the aggregate of history involves a `g.push`, and outgoing edges of changed nodes are going to touch a lot more nodes. In my implementation, I actually re-compute aggregate of history on demand during model forwarding, which makes the training 2-3x faster. @BarclayII

Reply:
the original example also computes the aggregation on demand. BTW, why do you guys not use the original implementation? Do you know DGL has a control variate implementation for more than one year?

Reply:
This control variate serves as an example for the new sampling API, I believe.

Reply:
it's pretty easy to adapt from the old implementation.

Reply:
Switched to implementation with blocks.  It is now 10 times faster.